[클린코드 4기 이상훈] 자판기 미션 STEP 1

sanghunlee-711 opened the pull request:
안녕하세요 윤호님! 자판기 미션 첫번째 스텝을 진행하였습니다. 

아래 간단한 코멘트 참고하셔서 리뷰해주시면 감사드리겠습니다 :)

2단계가 웹컴포넌트라는 것이 의식되어 2단계 진행 시 **재활용할 수 있는 로직이 많기를 바라는 마음**으로 최대한 분리하며 작성하였습니다.

또한, 아래 부분을 치중하여 진행하였습니다.

## MVC 패턴으로 가장 심플하게 구현해보기

* Model : 상태와 상태를 set하는 부분
* View: HTML에 해당하는 부분과 이를 변경하는 것, 이벤트바인딩
* Controller: Model과 View를 제어하여 원하는 앱이 구현 되도록 비즈니스로직을 포함.

이 과정에서 상단 탭을 담당하는 MenuView의 경우 다른 부분의 View까지 핸들링하게 되어 조금 아쉬운 부분이 있습니다.

태그의 형태로 보았을 때 과제에서 의도한건 해쉬아이디를 활용한 라우팅을 원하는 것 같다는 생각이 드는데 어떻게 생각하시는지 의견 여쭙고 싶습니다.

## 랜덤으로 동전생성

무작위로 충전이라는 말이 잘 이해가 가지 않아 아래와 같이 진행하였습니다.

* 금액이 입력된다
* 500, 100, 50 순으로 입력된 금액이 해당 값보다 큰 경우 랜덤한 값을 받아 각 코인을 카운팅 해주고 총 값을 빼준다
* 남은 값이 있는 경우 모두 10원으로 바꿔준다.
* 기존 값이 있는경우 카운팅 된 코인을 더해주고 `coinMap`이라는 state를 업데이트해준다.

혹시나 요구사항과 조금 동떨어지는 부분이라면 설명한번 부탁드리겠습니다!

Commit:
[Add] Install Cypress and jest

Commit:
[Add] HashRouting feature

Commit:
[Add] web storage handling feature

Commit:
[Add] Prettier, eslint

Commit:
[Add] Basic form for making app
- Trying to use MVC pattern for reusing logic in step2

Commit:
[Add] test case for step 1

Commit:
[Add] make class for dividing logic and view

Commit:
[Modify] change model name

Commit:
[Add] test code for typing product information

Commit:
[Add] Mange Product menu test code

Commit:
[Add] append product logic

Commit:
[Refactor] Remove Event class
- move event binding logic into view class

Commit:
[Add] Menu Logic

Commit:
[Add] charging coins test code

Commit:
[Add] reload logic with test code

Commit:
[Refactor] naming convention

Commit:
[Refactor] make commads in test code

yujo11 commented on .prettierrc.json:
```diff
@@ -0,0 +1,4 @@
+{
+  "printWidth": 120,
+  "singleQuote": true
+}
```
`prettier`의 default option들도 이것저것 설정 파일에 넣는 분들이 많던데 딱 필요한 것들만 넣으셨군요! 👍 👍 👍

yujo11 commented on REQUIREMENT.md:
```diff
@@ -0,0 +1,43 @@
+- 최초 상품 목록은 비워진 상태이다. (이미지 첨부)
+- 상품명, 금액, 수량을 추가할 수 있다.
+- 상품 추가 입력 폼에 상품명, 금액, 수량을 차례로 입력한다.
+- 상품명, 금액, 수량은 공백이 불가능하다.
+- 상품의 최소 수량은 1개여야 한다.
+- 상품의 최소 가격은 100원이며, 10원으로 나누어 떨어져야 한다.
+  - 예) 콜라 / 110원 / 5개
+  - 예) 사이다 / 100원 / 100개
+- 같은 상품명의 데이터를 추가하면 기존의 상품에 해당하는 데이터는 새로운 상품으로 대체된다.
+- 콜라 / 1000원 / 12개(전) -> 콜라 / 1500원 / 10개(후) => 콜라 / 1500원 / 10개(결과)
+- 사용자는 추가한 상품을 확인할 수 있다.
+- 상품의 이름, 가격, 수량 순으로 상품 목록이 보여진다. (이미지 첨부)
+- 상품 목록은 탭을 이동하여도 기존의 상품 목록이 유지되어야 한다.
+
+## 2. 잔돈 충전 / 요구사항
+
+- 잔돈 충전 탭에서, 다음과 같은 규칙으로 자판기 보유 금액을 충전한다.
+- 잔돈 충전 페이지에서 최초 자판기가 보유한 금액은 0원이며, 각 동전의 개수는 0개이다.
+- 관리자는 잔돈 충전 입력 요소에 충전할 금액을 입력한 후, 자판기 동전 충전 버튼을 눌러 자판기가 보유한 금액을 충전할 수 있다.
+- 최소 충전 금액은 100원이며, 10원으로 나누어 떨어지는 금액만 충전이 가능하다.
+- 자판기가 보유한 금액은 {금액}원 형식으로 나타낸다.
+  - 예) 1000원 (o) / 1000 원 (x) / 1000 (x)
+- 관리자는 잔돈을 누적하여 충전할 수 있다.
+- 1000원 충전 -> 500원 충전 => 총 1500원 분량의 동전이 생성됨. (추가)
+- 자판기가 보유한 금액 만큼의 동전이 무작위로 생성된다.
+- 동전은 500원, 100원, 50원, 10원의 동전만 생성된다.
+- 동전의 개수를 나타내는 정보는 {개수}개 형식으로 나타낸다.
+  - 예) 1개 (o) / 1 개 (x) / 1 (x)
+- 다른 탭을 클릭하여도 자판기가 보유한 금액은 유지되어야 한다.
+
+## 공통 요구사항
+
+- 기존 HTML Template을 활용한다.
+- 필요하다면 선택자를 참고한다.
+- 상단에 탭메뉴가 존재하며 각 탭에 따라 적절한 기능을 수행한다.
+- 상품 관리탭은 자판기가 보유하고 있는 물품을 추가하는 기능을 수행한다.
+- 잔돈 충전탭은 자판기가 보유할 금액을 충전하는 기능을 수행한다.
+- 상품 구매탭은 사용자가 금액을 충전할 수 있으며, 금액에 맞춰 상품을 구매하고, 남은 금액에 대해서는 잔돈을 반환하는 기능을 수행한다.
+- 다른 탭으로 이동했다 돌아와도 기존 탭의 상태가 유지되어야 한다.
+- localStorage를 이용하여, 새로고침하더라도 가장 최근에 작업한 정보들을 불러올 수 있도록 한다.
```
요구사항을 정리해놓으신 부분이 좋네용 👍 👍 👍

sanghunlee-711 replied:
@yujo11  매번까먹는게 생겨서 그냥 복사 붙여넣기로 하나 만들어 버렸습니다 ㅎㅎ

yujo11 commented on cypress/e2e/vending-machine.cy.js:
```diff
@@ -0,0 +1,253 @@
+  });
+});
+
+describe('잔돈 충전하기', () => {
+  beforeEach(() => {
+    cy.visit('../../index.html');
+    cy.get($ELEMENT.VENDING_MACHINE_MANANGE_MENU).click();
+  });
+  context('잔돈 입력 시', () => {
+    beforeEach(() => {
+      cy.clearLocalStorage();
+    });
+
+    it('충전 값을 입력할 input이 존재해야 한다.', () => {
+      cy.get($ELEMENT.CHARGE_INPUT).should('exist');
+    });
```
`describe`, `context`, `it`을 의미에 맞게 잘 활용하고 계시네용 👍 👍 👍

yujo11 commented on index.html:
```diff
@@ -1,18 +1,86 @@
+	<button id="vending-machine-manage-menu" class="hash-nav">잔돈충전</button>
+	<!-- <button id="product-purchase-menu" class="hash-nav">상품 구매</button> -->
+
+	<div id="appending-product-container" class="contents-container">
+		<h3>상품 추가하기</h3>
+		<div class="product-container">
```
이건 그냥 궁금한건데 4space를 사용하시나용?

sanghunlee-711 replied:
![image](https://user-images.githubusercontent.com/57477415/209071137-f0c9f2df-9446-4520-b124-c6d5f721de46.png)

2를 사용하고 있는데 코드에 왜 저렇게 나타나는지 허허..
에디터에서는 하단과 같이 잘 나타나고 있습니다 ;-;..
![image](https://user-images.githubusercontent.com/57477415/209071239-bcb203b3-88d2-40db-ad27-a95604ec0f26.png)

yujo11 commented on jest.config.js:
```diff
@@ -0,0 +1,19 @@
+/** @type {import('jest').Config} */
+const config = {
+  moduleFileExtensions: ['js', 'json', 'jsx', 'ts', 'tsx', 'json'],
+  transform: {
+    '^.+\\.(js|jsx)?$': 'babel-jest',
+  },
+  testEnvironment: 'node',
+  moduleNameMapper: {
+    '^@/(.*)$': '<rootDir>/$1',
+  },
+  testMatch: [
+    '<rootDir>/**/*.test.(js|jsx|ts|tsx)',
+    '<rootDir>/(tests/unit/**/*.spec.(js|jsx|ts|tsx)|**/__tests__/*.(js|jsx|ts|tsx))',
+  ],
+  transformIgnorePatterns: ['<rootDir>/node_modules/'],
+  verbose: true,
+};
+
+module.exports = config;
```
Jest를 사용해서 unit test를 작성하려고 하셨던걸까요?

sanghunlee-711 replied:
@yujo11  네 맞습니다 복잡한 함수의 경우 별도로 테스트를 진행해보려했는데.. 작성을 하나도 안했네요 😅 스텝2에서는 아마 할 수 있을지도.. 라고 희망을 가져봅니다

yujo11 commented on src/constants/alert.js:
```diff
@@ -0,0 +1,4 @@
+export const ALERT = {
+  PRICE_VALIDATION: '유효하지 않은 가격입니다. 상품의 최소가격은 100원이며 10원으로 나누어 떨어져야 합니다.',
+  CHARGE_VALIDATION: '유효하지 않은 충전 가격입니다. 충전 최소금액은 100원이며 10원으로 나누어 떨어져야 합니다.',
+};
```
100원과 10원에 들어가는 숫자들은 정책에 따라 변경이 일어날 수 있는 부분인데 별도로 상수화 해서  사용하셔도 좋을거 같네요

sanghunlee-711 replied:
@yujo11 좋은 생각인 것 같아요!

yujo11 commented on src/constants/common.js:
```diff
@@ -0,0 +1,25 @@
+  typedCoin: 0,
+  chargedTotal: 0,
+  coinMap: DEFAULT_CHARGE_COINS,
+};
```
이 값들은 상태를 초기화하기 위해 사용되는거 같은데 `common`보다는 분리해놓으신 다른 상수 모듈들처럼 의미를 더 드러낼 수 있게 모듈이름을 바꿔주셔도 좋을거 같아요

yujo11 commented on src/constants/element.js:
```diff
@@ -0,0 +1,22 @@
+export const $ELEMENT = {
+  NAME_INPUT: '#product-name-input',
+  PRICE_INPUT: '#product-price-input',
+  QUANTITY_INPUT: '#product-quantity-input',
+  ADD_BUTTON: '#product-add-button',
+  INVENTORY_CONTAINER: '#product-inventory-container',
+  APPENDING_CONTAINER: '#appending-product-container',
+  CHARGING_CONTAINER: '#charging-money-container',
+  PRODUCT_LIST: '.product-inventory-list',
+  MENU: '.hash-nav',
+  PRODUCT_MANGNE_MENU: '#product-manage-menu',
+  VENDING_MACHINE_MANANGE_MENU: '#vending-machine-manage-menu',
+  PRODUCT_PURCHASE_MENU: '#product-purchase-menu',
+  CHARGE_INPUT: '#charge-input',
+  CHARGE_BUTTON: '#charge-button',
+  CHARGE_AMOUNT: '#charge-amount',
+  CASH_BOX_CONTAINER: '.cashbox-change',
+  COIN_500_COUNT: '#coin-500-quantity',
+  COIN_100_COUNT: '#coin-100-quantity',
+  COIN_50_COUNT: '#coin-50-quantity',
+  COIN_10_COUNT: '#coin-10-quantity',
+};
```
이 부분은 취향일 수도 있는데 중간중간 카테고리 별로 개행이 들어가면 한 눈에 코드를 보기 좋을거 같아요

sanghunlee-711 replied:
@yujo11 좋은 것 같습니다 아예 하나의 depth를 더 넣어서 나눠야하나라는 생각도 했었는데 번거로움이 큰 것 같았습니다 ㅎㅎ..
이야기해주신 것 처럼 차라리 개행으로 간편하게 구분해보는게 좋을 것 같아요 👍

yujo11 commented on src/js/controller/chargeController.js:
```diff
@@ -0,0 +1,69 @@
+    const coin10 = coins >= 10 ? Math.floor(coins / 10) : 0;
+    coins -= coin10 * 10;
+
+    return {
+      coin500,
+      coin100,
+      coin50,
+      coin10,
+    };
```
👍👍👍

yujo11 commented on src/js/controller/productController.js:
```diff
@@ -0,0 +1,121 @@
+    return isValidPrice;
+  };
+
+  checkSameProduct = ({ products, typedProduct }) => {
+    let isModified = false;
+
+    const newProducts = products.map((product) => {
+      const { name } = product;
+      if (name === typedProduct.name) {
+        isModified = true;
+        return {
+          name,
+          price: typedProduct.price,
+          quantity: typedProduct.quantity,
+        };
+      } else {
+        return product;
+      }
+    });
+
+    if (!isModified) return [...newProducts, typedProduct];
+
+    return newProducts;
+  };
```
flag 역할을 하는 `isModified`가 사용 되고, depth가 생김에 따라 코드를 직관적으로 이해하기 어려웠는데요. 

조금 더 읽기 쉽게 작성하는 방법은 없을까요?

sanghunlee-711 replied:
이 부분은 저도 작성하면서 아쉬움이 있던 부분입니다! 조금 머리를 더 써보겠습니다 :)

yujo11 commented on src/js/controller/productController.js:
```diff
@@ -0,0 +1,121 @@
+  };
+
+  validatePrice = ({ price }) => {
+    const isValidPrice = price % 10 === 0 && price >= VALIDATE.MIN_PRICE;
```
10과 같은 매직 넘버도 상수화가 되면 좋을거 같아요

yujo11 commented on src/js/view/chargeView.js:
```diff
@@ -0,0 +1,44 @@
+
+      if (event.key === 'Enter') {
+        addCharge();
+      }
```
Enter 키 입력으로 사용성을 생각해주신 부분이 좋네요 👍 👍

yujo11 commented on src/js/view/menuView.js:
```diff
@@ -0,0 +1,39 @@
+    this.renderTab({ tabId: VENDING_MACHINE_INITIAL_STATE.hashId });
+  }
+
+  renderTab = ({ tabId }) => {
+    const { $menus, MENU_CONTAINER_MAP } = this;
+
+    $menus.forEach((menu) => {
+      if (menu.id === tabId) {
+        menu.classList.add('active');
+        MENU_CONTAINER_MAP.get(menu.id).classList.remove('invisible');
+      } else {
+        menu.classList.remove('active');
+        MENU_CONTAINER_MAP.get(menu.id).classList.add('invisible');
+      }
+    });
+  };
```
이 부분은 이후 N개의 탭이 생기는 상황을 고려해서 이렇게 작성해주신걸까용?

sanghunlee-711 replied:
이야기해주신대로 나름 확장성을 고려한 시도였습니다! 그래서 별도의 클래스(.hash-nav)를 두고 querySelectorAll을 사용하였습니다 :)

yujo11 commented on index.html:
```diff
@@ -1,18 +1,86 @@
+		<h3>상품 추가하기</h3>
+		<div class="product-container">
+			<input type="text" id="product-name-input" placeholder="상품명" />
+			<input type="number" id="product-price-input" placeholder="가격" />
```
소소하게 최소 입력단위가 10원인만큼 `step` atrribute가 추가되면 좋을거 같네요

sanghunlee-711 replied:
@yujo11 오.. step이라는 속성을 코멘트를 통해 처음 봤습니다 좋은 속성이네요 적용해보겠습니다!

yujo11 commented on index.html:
```diff
@@ -1,18 +1,86 @@
+		<div class="purchase-container">
+			<h3>충전하기</h3>
+			<div class="vending-machine-wrapper">
+				<input type="number" name="charge-amount" id="charge-input" placeholder="충전할 금액을 입력해주세요" />
```
여기두요!

yujo11 commented on index.html:
```diff
@@ -1,18 +1,86 @@
+	<div id="charging-money-container" class="contents-container">
+		<div class="purchase-container">
+			<h3>충전하기</h3>
+			<div class="vending-machine-wrapper">
+				<input type="number" name="charge-amount" id="charge-input" placeholder="충전할 금액을 입력해주세요" />
+				<button id="charge-button">충전하기</button>
+			</div>
```
현재 `keyup` 이벤트에서 `Enter`키 입력을 잡아서 핸들링하고 계신데 `input`들을 `form`으로 감싸서 `submit`이벤트를 핸들링하도록 변경해보시는건 어떨까요?

sanghunlee-711 replied:
@yujo11 좋은 것 같습니다! html을 더욱 의미있게 작성하게 되는 방향인 것 같아요!
 반영해보겠습니다 :)

yujo11 reviewed (changes requested):
안녕하세요 상훈님!

코드가 전체적으로 굉장히 깔끔하게 작성되어 있어서 코드를 보면서 오히려 제가 많이 배워갔네요

제공되는 CSS가 바뀐건지 LMS에 예시로 첨부된 자판기 앱보다 UI도 예뻐서 좋았네요 ㅋㅋ

코멘트 드릴 내용이 많지 않아 몇 가지 간단한 의견만 남겼는데요. 확인 해주시면 감사하겠습니다!

---

>태그의 형태로 보았을 때 과제에서 의도한건 해쉬아이디를 활용한 라우팅을 원하는 것 같다는 생각이 드는데 어떻게 생각하시는지 의견 여쭙고 싶습니다.

이 부분은 제가 이 과정을 설계한게 아니기 때문에 뭐라고 확답을 드리기 힘드네요

해쉬아이디를 활용한 라우팅이 Hash Router를 말씀하시는거라면 구현 과정에서 차이를 찾아보게 될테니 어떤 방식으로 구현해도 상관 없을거 같다는 생각이 드네요

>무작위로 충전이라는 말이 잘 이해가 가지 않아 아래와 같이 진행하였습니다.

아래에 적어주신 로직을 보면 충분히 납득 가도록 구현하신거 같습니다. 

이 부분은 저도 에매하다고 생각해서 따로 질문 남겼는데 답변을 받는다면 다시 공유 드리도록 하겠습니다

---

미션을 빠르게 제출해주셨는데 리뷰 남기는게 조금 늦었네요 고생 많으셨습니다 상훈님!! 👍 👍 👍 👍

Commit:
[Refactor] Apply reviews
- dividing common constant file to each meaningful files
- using html attributes
- removing magic number hard coded by making constant

Commit:
[Refactor] productController > checkSameProduct with test code

sanghunlee-711 commented:
> 안녕하세요 상훈님!
> 
> 코드가 전체적으로 굉장히 깔끔하게 작성되어 있어서 코드를 보면서 오히려 제가 많이 배워갔네요
> 
> 제공되는 CSS가 바뀐건지 LMS에 예시로 첨부된 자판기 앱보다 UI도 예뻐서 좋았네요 ㅋㅋ
> 
> 코멘트 드릴 내용이 많지 않아 몇 가지 간단한 의견만 남겼는데요. 확인 해주시면 감사하겠습니다!
> 
> > 태그의 형태로 보았을 때 과제에서 의도한건 해쉬아이디를 활용한 라우팅을 원하는 것 같다는 생각이 드는데 어떻게 생각하시는지 의견 여쭙고 싶습니다.
> 
> 이 부분은 제가 이 과정을 설계한게 아니기 때문에 뭐라고 확답을 드리기 힘드네요
> 
> 해쉬아이디를 활용한 라우팅이 Hash Router를 말씀하시는거라면 구현 과정에서 차이를 찾아보게 될테니 어떤 방식으로 구현해도 상관 없을거 같다는 생각이 드네요
> 
> > 무작위로 충전이라는 말이 잘 이해가 가지 않아 아래와 같이 진행하였습니다.
> 
> 아래에 적어주신 로직을 보면 충분히 납득 가도록 구현하신거 같습니다.
> 
> 이 부분은 저도 에매하다고 생각해서 따로 질문 남겼는데 답변을 받는다면 다시 공유 드리도록 하겠습니다
> 
> 미션을 빠르게 제출해주셨는데 리뷰 남기는게 조금 늦었네요 고생 많으셨습니다 상훈님!! 👍 👍 👍 👍

우선 칭찬을 해주셔서 격렬한 감사 드립니다 😭😭😭. 아마도 스터디 과정중에 늘긴 늘었나봐요!.. 
이야기 해주신 코멘트 기반으로 리팩토링을 진행했습니다.
## html의 속성 사용하기
- input의 step, form의 submit이벤트 활용

## 파일이름을 통해 분리하기
* common.js -> initialState.js, storage.js등으로 상수를 조금 더 명확하게 분리했습니다

## 매직넘버 줄이기
* 별도의 상수로 뺀 다음 alert.js에 적용되는 문구에도 활용하게 되니 추후 수정을 생각한다면 훨씬 용이해진 것 같습니다

## 가독성이 떨어지는 로직을 변경해보기
* 나름 변경한다고 해보았으나 .. 조금 더 좋은 방법이 있을 것 같은데 마땅히 떠오르는게 없습니다 ㅎㅎ;;

## Jest를 통한 간단한 테스팅 진행
위 가독성이 떨어지는 로직을 변경하기 전 간단하게 Controller만 테스트를 진행해보았습니다.

DOM이 필요하게 되어 testHtml.js라는 파일에 body에 해당하는 태그를 모두 넣고 테스트 전에 innerHTML을 실행하게 되었습니다.

이를 더 우아하게 해결할 방법은 딱히 못찾겠더라구요 🥲..

혹시 더 좋은 방법이 보이신다면 조언부탁드리겠습니다!

yujo11 commented on index.html:
```diff
@@ -41,10 +41,10 @@ <h3>상품 추가하기</h3>
 			<h3>충전하기</h3>
-			<div class="vending-machine-wrapper">
-				<input type="number" name="charge-amount" id="charge-input" placeholder="충전할 금액을 입력해주세요" />
+			<form class="vending-machine-wrapper">
+				<input type="number" name="charge-amount" id="charge-input" placeholder="충전할 금액을 입력해주세요" step="10" min="0" />
```
`<form>` 태그와 `step` attribute를 사용해주셨군요!  👍👍👍

yujo11 commented on src/constants/alert.js:
```diff
@@ -1,4 +1,6 @@
 export const ALERT = {
-  PRICE_VALIDATION: '유효하지 않은 가격입니다. 상품의 최소가격은 100원이며 10원으로 나누어 떨어져야 합니다.',
-  CHARGE_VALIDATION: '유효하지 않은 충전 가격입니다. 충전 최소금액은 100원이며 10원으로 나누어 떨어져야 합니다.',
+  PRICE_VALIDATION: `유효하지 않은 가격입니다. 상품의 최소가격은 ${VALIDATE.MIN_PRICE}원이며 ${VALIDATE.MIN_UNIT}원으로 나누어 떨어져야 합니다.`,
+  CHARGE_VALIDATION: `유효하지 않은 충전 가격입니다. 충전 최소금액은 ${VALIDATE.MIN_PRICE}원이며 ${VALIDATE.MIN_UNIT}원으로 나누어 떨어져야 합니다.`,
```
👍👍👍

yujo11 commented on src/constants/element.js:
```diff
@@ -3,17 +3,22 @@ export const $ELEMENT = {
   PRICE_INPUT: '#product-price-input',
   QUANTITY_INPUT: '#product-quantity-input',
   ADD_BUTTON: '#product-add-button',
+  PRODUCT_LIST: '.product-inventory-list',
+
   INVENTORY_CONTAINER: '#product-inventory-container',
   APPENDING_CONTAINER: '#appending-product-container',
   CHARGING_CONTAINER: '#charging-money-container',
-  PRODUCT_LIST: '.product-inventory-list',
+
   MENU: '.hash-nav',
   PRODUCT_MANGNE_MENU: '#product-manage-menu',
   VENDING_MACHINE_MANANGE_MENU: '#vending-machine-manage-menu',
   PRODUCT_PURCHASE_MENU: '#product-purchase-menu',
+
+  CHARGE_CONTAINER: '.vending-machine-wrapper',
   CHARGE_INPUT: '#charge-input',
   CHARGE_BUTTON: '#charge-button',
   CHARGE_AMOUNT: '#charge-amount',
+
   CASH_BOX_CONTAINER: '.cashbox-change',
   COIN_500_COUNT: '#coin-500-quantity',
```
개행이 들어가면서 코드 보기가 좋아졌네요!

yujo11 commented on src/constants/validate.js:
```diff
@@ -1,5 +1,6 @@
 export const VALIDATE = {
   MIN_PRICE: 100,
+  MIN_UNIT: 10,
```
👍👍👍

yujo11 commented on src/js/controller/__test__/productController.test.js:
```diff
@@ -0,0 +1,47 @@
+import ProductView from '../../view/productView.js';
+import ProductController from '../productController.js';
+
+describe('productController', () => {
```
unit test도 작성해주셨군요! 👍 👍 👍

yujo11 replied:
그런데 이 부분만 `cypress`로 테스트하지 않고 Jest를 사용해서 별도로 작성해 주셨는데 이유가 있을까요?

sanghunlee-711 replied:
@yujo11 해당 부분만 테스트를 어떻게 진행하면 좋을까.. 고민하다가 Cypress로 진행하는 것 보다 조금 더 개인적으로 익숙한 Jest로 진행하게 되었습니다.

cypress에서 productController를 처음 생성하는 것에 에러가 계속 발생하여 Jest를 사용해보았습니다.

아무래도 제가 stub과 같은 기능을 저번 과제에서도 결국 적용을 못하고 있는 상태라 불신이 커서 그런가 봅니다 ㅎㅎ...

다음 스텝에서 테스트 코드를 한번 변경하면서 적용해보겠습니다!

yujo11 commented on src/js/controller/productController.js:
```diff
@@ -28,33 +28,21 @@ class ProductController {
-    });
+    const isExist = products.findIndex((product) => product.name === typedProduct.name) > -1;
 
-    if (!isModified) return [...newProducts, typedProduct];
+    if (isExist) {
+      return products.map((product) => (product.name === typedProduct.name ? typedProduct : product));
+    }
 
-    return newProducts;
+    return [...products, typedProduct];
```
리팩터링 하는게 쉽지 않은 작업이었을텐데 리팩터링을 통해 이전보다 훨씬 직관적으로 코드가 변했네요! 멋집니다!! 👍 👍 👍

sanghunlee-711 replied:
전보다 보기 좋다니 다행입니다 🥲

yujo11 commented on src/js/view/chargeView.js:
```diff
@@ -27,18 +28,20 @@ export default class ChargeView {
   };
 
-  addTypeChargeEvent = ({ typeCoin, addCharge }) => {
+  addSubmitChargeEvent = (addCharge) => {
+    this.$chargeContainer.addEventListener('submit', (event) => {
+      addCharge(event);
+    });
+  };
```
👍👍👍

yujo11 reviewed (approved):
안녕하세요 상훈님!

리뷰 남긴 내용을 빠르게 반영해주셨군요! 

기존에 코드가 깔끔하게 작성되어 있어서  변경된 코드를 보는 것도 어렵지 않았네요!

다만 개인적으로 코드를 커밋 단위로 보고는 하는데 리뷰 이후의 커밋이 큰 덩어리로 되어 있어서 각각의 변경점들을 함께 봐야 하는 부분이 약간은 불편하게 느껴졌네요. 커밋 단위를 조금 더 쪼개 주셔도 좋을거 같다는 생각입니다!

하나 궁금한 점은 코멘트로 남겨놓기도 했지만 Jest를 사용한 Unit Test를 작성하시면서 DOM을 목킹하는 등의 공수가 꽤나 많이 들었을거 같은데요. 

기존에 작성하셨던 Cypress 테스트 코드에 해당 Controller에 대한 테스트를 추가하면 이런 과정없이 다소 편하게(?) 테스트가 가능했을거 같은데 해당 Controller만 테스트를 별도로 작성하신 이유가 궁금하네용

---

이번 step은 여기에서 이만 merge해도 충분할거 같습니다!

상훈님이 작성하신 코드를 보면서 저도 배워가는 점이 많네요! 다음 step도 기다리고 있겠습니다! 

고생 많으셨습니다 상훈님!! 👍 👍 👍
